File: 백준/Silver/1931. 회의실 배정/회의실 배정.py
import sys

N = int(sys.stdin.readline())
arr = [list(map(int, sys.stdin.readline().split())) for _ in range(N)]

# 회의를 종료 시간 기준으로 오름차순 정렬. 종료 시간이 같으면 시작 시간으로 정렬
arr.sort(key=lambda x: (x[1], x[0]))

count = 0
current_end_time = 0

for i in range(N):
    # 현재 회의의 시작 시간이 이전에 선택한 회의의 종료 시간보다 크거나 같다면
    if arr[i][0] >= current_end_time:
        count += 1  # 회의를 선택
        current_end_time = arr[i][1]  # 선택한 회의의 종료 시간을 기록

print(count)

File: 백준/Silver/1931. 회의실 배정/회의실 배정.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Cop{
    private int start;
    private int end;

    public Cop(int start, int end) {
        this.start = start;
        this.end = end;
    }

    public int getStart(){
        return this.start;
    }

    public int getEnd(){
        return this.end;
    }
}

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        String s = bf.readLine();
        int N = Integer.parseInt(s);
        int result = 0;

        List<Cop> list = new ArrayList<>();

        for(int i=0; i<N; i++){
            String[] sa = bf.readLine().split(" ");
            list.add(new Cop(Integer.parseInt(sa[0]), Integer.parseInt(sa[1])));
        }

        Collections.sort(list, (o1, o2)->{
            // 끝나는 시간이 같으면
            if(o1.getEnd() == o2.getEnd()){
                return o1.getStart() - o2.getStart();
            }
            return o1.getEnd() - o2.getEnd();
        });

        int cur_end = 0;


        for(Cop c : list){
            if(cur_end <= c.getStart()){
                cur_end = c.getEnd();
                result++;
            }
        }

        System.out.println(result);
    }
}
